components/cldr-json-data-provider/src/support.rs

```diff
@@ -0,0 +1,69 @@
+use crate::CldrPaths;
+use icu_data_provider::iter::DataEntryCollection;
+use icu_data_provider::prelude::*;
+use std::convert::TryFrom;
+use std::sync::RwLock;
+
+pub(crate) trait DataKeySupport {
+    fn supports_key(data_key: &DataKey) -> Result<(), DataError>;
+}
+
+pub(crate) struct LazyCldrProvider<T> {
+    data_provider: RwLock<Option<T>>,
+}
+
+impl<'b, 'd, T> LazyCldrProvider<T>
+where
+    T: DataProvider<'d> + DataKeySupport + DataEntryCollection + TryFrom<&'b CldrPaths>,
+    <T as TryFrom<&'b CldrPaths>>::Error: 'static + std::error::Error,
+{
+    pub fn new() -> Self {
+        Self {
+            data_provider: RwLock::new(None),
```

Comment:
nit: the name `data_provider` here doesn't seem to match what seems like a file mutex type.

Reply:
@zbraniecki @Manishearth 

As a more general question, what I'm trying to do in this file is to lazy-create a data provider object (type `T`).  To do this, I used an `RwLock<Option<T>>`, defaulting to `None`, and then when it is needed, if it is `None`, I create it and store it into the field.

Is this the right pattern?  Can you suggest better ways to achieve the "lazy load" pattern?

Reply:
You didn't ask me :), but I have two cents.

Why is it important to have a lazy provider, versus always providing eagerly, and then asking the user to lazify the provision if they want to?  For example, the crate `lazy_init` can do that. https://docs.rs/lazy-init/0.3.0/lazy_init/